src/python/pants/backend/go/module.py

```diff
@@ -174,6 +179,44 @@ async def resolve_go_module(
     )
 
 
+@dataclass(frozen=True)
+class FindOwningGoModuleRequest:
+    address: Address
+
+
+@dataclass(frozen=True)
+class ResolvedOwningGoModule:
+    module_address: Optional[Address]
+
+
+@rule
+async def find_nearest_go_module(request: FindOwningGoModuleRequest) -> ResolvedOwningGoModule:
+    # Obtain unexpanded targets and ensure file targets are filtered out. Unlike Python, file targets do not
+    # make sense semantically for Go source since Go builds entire packages at a time. The filtering is
```

Comment:
Just a note: this is somewhat true of Java as well: the entire JVM package is implicitly available without an import statement, but separate compilation is still supported.

I expect that @patricklaw will have the inference implementation for the JVM infer a dependency on all files in the package (which will cause it to automatically be coarsened to a single `CoarsenedTarget`) rather than taking the approach you're taking here. Part of the reason for that is that the JVM allows import cycles between files, and so he already needs to be able to support cycles and thus coarsening.

The equivalent case for `go` would be if multiple packages had cyclic imports for one another: but it looks like that  isn't possible: https://jogendra.dev/import-cycles-in-golang-and-how-to-deal-with-them

So yea: this seems like the way to go for `go`.